src/lib.rs

```diff
@@ -119,6 +117,63 @@ mod imports {
 
 use imports::*;
 
+/// A trait representing the implementation behavior of an arena and how
+/// identifiers are represented.
+///
+/// ## When should I implement `ArenaBehavior` myself?
+///
+/// Usually, you should just use `DefaultArenaBehavior`, which is simple and
+/// correct. However, there are some scenarios where you might want to implement
+/// `ArenaBehavior` yourself:
+///
+/// * **Space optimizations:** The default identifier is two words in size,
+/// which is larger than is usually necessary. For example, if you know that an
+/// arena *cannot* contain more than 256 items, you could make your own
+/// identifier type that stores the index as a `u8` and then you can save some
+/// space.
+///
+/// * **Trait Coherence:** If you need to implement an upstream crate's traits
+/// for identifiers, then defining your own identifier type allows you to work
+/// with trait coherence rules.
+///
+/// * **Share identifiers across arenas:** You can coordinate and share
+/// identifiers across different arenas to enable a "struct of arrays" style
+/// data representation.
+pub trait ArenaBehavior {
+    /// The identifier type.
+    type Id: Copy;
+
+    /// Construct a new object identifier from the given index and arena
+    /// identifier.
+    ///
+    /// ## Panics
+    ///
+    /// Implementations are allowed to panic if the given index is larger than
+    /// the underlying storage (e.g. the implementation uses a `u8` for storing
+    /// indices and the given index value is larger than 255).
+    fn new_id(index: usize, arena_id: usize) -> Self::Id;
```

Comment:
The order of arguments feels "swapped" here: arena is "big", I'd expect it to be first.